### DISCUSSION TITLE
Think about how to pull patterns from GPU devices

### DISCUSSION THREAD

**Discussion by jyoung3131** [hpcgarage/spatter#176] [Ideas]
Right now gs_patterns uses Intel PIN to pull CPU traces and bin them into appropriate Spatter inputs. This workflow works well, but it does not provide a true "fair" comparison for GPU Spatter benchmarking runs since the inputs are targeted towards a CPU cache-based architecture. To provide a better comparison, we should use an Intel PIN equivalent that traces directly from GPU applications (CUDA/HIP/OneAPI). 

Our understanding is that [NVBit](https://github.com/NVlabs/NVBit) could be used in a similar way as Intel Pin or DynamoRio to pull traces from CUDA applications that could be used as inputs to the gs_patterns tool. This is likely a large project that would require significant effort, but we should consider and flesh out the steps to do so. 

One noted caveat is that NVBit does not currently support H100 GPUs (sm_90), so we'd need to test with A100 or earlier.

**Comment by plavin** [comment]
This feature has been added to LANL's GS Patterns tool. Here is a link to [the HPC Garage fork](https://github.com/hpcgarage/gs_patterns/tree/nvbit_support), which will hopefully be merged upstream soon.
